From 7ff9361f72e8d905cf0373bbd4e9e06e3e3215e4 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Date: Wed, 23 Apr 2025 16:40:06 +0200
Subject: [PATCH 1675/1695] drivers: stm32_risaf: fix HWCFGR information
 management

hw_nregions (and so hw_nsubregions) take into account the base0 region,
which is not configurable. So decrement their value after read.
As described in the reference manual, hw_nsubregions reflects the total
number of subregions A and B. Convert it to the number of subregions per
region, easier to use from software point of  view.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2e6ae0c748b21aebcb76dd952db1863c3eede7b0 drivers: stm32_risaf: fix HWCFGR information management]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Change-Id: I7dd77aa9f40cedb82eb37fb86e454666f7d25d08
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/454255
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Tested-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/drivers/firewall/stm32_risaf.c | 20 ++++++++++++++------
 1 file changed, 14 insertions(+), 6 deletions(-)

diff --git a/core/drivers/firewall/stm32_risaf.c b/core/drivers/firewall/stm32_risaf.c
index 5fbacd590..f9c3df195 100644
--- a/core/drivers/firewall/stm32_risaf.c
+++ b/core/drivers/firewall/stm32_risaf.c
@@ -673,16 +673,24 @@ static TEE_Result stm32_risaf_init_ddata(struct stm32_risaf_instance *risaf)
 	mask_msb = mask_lsb + ((hwcfgr & _RISAF_HWCFGR_CFG4_MASK) >>
 			       _RISAF_HWCFGR_CFG4_SHIFT) - 1U;
 	risaf->ddata->mask_regions = GENMASK_32(mask_msb, mask_lsb);
-	risaf->ddata->max_base_regions = (hwcfgr & _RISAF_HWCFGR_CFG1_MASK) >>
-					 _RISAF_HWCFGR_CFG1_SHIFT;
+	/* hw_nregions take account the base0, which is not configurable */
+	risaf->ddata->max_base_regions = ((hwcfgr & _RISAF_HWCFGR_CFG1_MASK) >>
+					  _RISAF_HWCFGR_CFG1_SHIFT) - 1;
 
 	/* Get IP region granularity */
 	risaf->ddata->granularity = BIT((hwcfgr & _RISAF_HWCFGR_CFG3_MASK) >>
 					_RISAF_HWCFGR_CFG3_SHIFT);
 
-	/* Get maximum number of subregions */
-	risaf->ddata->max_subregions = (hwcfgr & _RISAF_HWCFGR_CFG2_MASK) >>
-					_RISAF_HWCFGR_CFG2_SHIFT;
+	/*
+	 * Get maximum number of subregions.
+	 * hw_nsubregions reflects the total number of subregions A and B.
+	 * It also takes into account the base0, decrement it.
+	 * Convert it to the number of subregions per region.
+	 */
+	mask_lsb = ((hwcfgr & _RISAF_HWCFGR_CFG2_MASK) >>
+		   _RISAF_HWCFGR_CFG2_SHIFT) - 1;
+	risaf->ddata->max_subregions = (mask_lsb * 2) /
+				       risaf->ddata->max_base_regions;
 
 	return TEE_SUCCESS;
 }
@@ -1224,7 +1232,7 @@ static TEE_Result stm32_risaf_probe(const void *fdt, int node,
 			panic();
 
 		id = _RISAF_GET_REGION_ID(regions[i].cfg);
-		assert(id < risaf->ddata->max_base_regions);
+		assert(id <= risaf->ddata->max_base_regions);
 
 		if (risaf_configure_region(risaf, &regions[i]))
 			panic();
-- 
2.47.2

